src/ConnectedMode/Persistence/ServerConnectionsRepository.cs

```diff
@@ -0,0 +1,207 @@
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Core.Binding;
+using System.ComponentModel.Composition;
+using System.IO;
+using SonarLint.VisualStudio.Core.Persistence;
+
+namespace SonarLint.VisualStudio.ConnectedMode.Persistence;
+
+
+[Export(typeof(IServerConnectionsRepository))]
+[PartCreationPolicy(CreationPolicy.Shared)]
+internal class ServerConnectionsRepository : IServerConnectionsRepository
+{
+    private readonly ISolutionBindingCredentialsLoader credentialsLoader;
+    private readonly ILogger logger;
+    private readonly IJsonFileHandler jsonFileHandle;
+    private readonly IServerConnectionModelMapper serverConnectionModelMapper;
+    private readonly string storageFilePath;
+    private const string ConnectionsFileName = "connections.json";
+
+    private ConcurrentDictionary<string, ServerConnection> ServerConnectionsCache { get; set; }
+
+    [ImportingConstructor]
+    public ServerConnectionsRepository(
+        IJsonFileHandler jsonFileHandle,
+        IServerConnectionModelMapper serverConnectionModelMapper,
+        ISolutionBindingCredentialsLoader credentialsLoader,
+        ILogger logger) : this(jsonFileHandle,
+        serverConnectionModelMapper,
+        credentialsLoader,
+        EnvironmentVariableProvider.Instance,
+        logger) { }
+
+    internal /* for testing */ ServerConnectionsRepository(
+        IJsonFileHandler jsonFileHandle,
+        IServerConnectionModelMapper serverConnectionModelMapper,
+        ISolutionBindingCredentialsLoader credentialsLoader,
+        IEnvironmentVariableProvider environmentVariables,
+        ILogger logger)
+    {
+        this.jsonFileHandle = jsonFileHandle;
+        this.serverConnectionModelMapper = serverConnectionModelMapper;
+        this.credentialsLoader = credentialsLoader;
+        this.logger = logger;
+        storageFilePath = GetStorageFilePath(environmentVariables);
+    }
+
+    public bool TryGet(string connectionId, out ServerConnection serverConnection)
+    {
+        serverConnection = null;
+        if (ServerConnectionsCache == null)
+        {
+            ReadServerConnectionsFile();
+        }
+
+        if (ServerConnectionsCache != null && ServerConnectionsCache.TryGetValue(connectionId, out ServerConnection foundConnection))
+        {
+            serverConnection = foundConnection;
+            serverConnection.Credentials = credentialsLoader.Load(serverConnection.CredentialsUri);
+            return true;
+        }
+
+        return false;
+
+    }
+
+    public IReadOnlyList<ServerConnection> GetAll()
+    {
+        if (ServerConnectionsCache == null)
+        {
+            ReadServerConnectionsFile();
+        }
+
+        return ServerConnectionsCache?.Values.ToList() ?? [];
+    }
+
+    public bool TryAdd(ServerConnection connection)
+    {
+        return SafeUpdateConnectionsFile(() =>
+        {
+            var wasAdded = ServerConnectionsCache.TryAdd(connection.Id, connection);
```

Comment:
All connections added here will keep holding the credentials reference. Not sure if we want that.

Reply:
What do you mean? The connections added here have to have the credentials. There should never be a connection added without credentials.

Reply:
The ones that are read from disk won't have credentials loaded

Reply:
The credentials have to be passed to be able to store them. So if they are filled, you want them to be removed when calling TryAdd? That would be a very unexpected behaviour.

Furthermore, the credentials are also filled when calling TryGet, but not when GetAll. I do not think we should change anything here.

Reply:
Yeah I'm not sure what we need to do, but I find it equally confusing that on GetAll _some_ connections have credentials and some don't

Reply:
I know. You said in the daily: we should do it how it was done in the SolutionBindingRepository. 

Anyway, at this point, the ServerConnection object has already credentials filled. So removing them is totally unexpected behavior.